// botfuel-dialog/build/nlus/botfuel-nlu.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _get = function get(object, property, receiver) { if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * Copyright (c) 2017 - present, Botfuel (https://www.botfuel.io).
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

var fs = require('fs');
var fsExtra = require('fs-extra');
var rp = require('request-promise-native');
var dir = require('node-dir');
var logger = require('logtown')('BotfuelNlu');
var BooleanExtractor = require('../extractors/boolean-extractor');
var LocationExtractor = require('../extractors/location-extractor');
var CompositeExtractor = require('../extractors/composite-extractor');
var SdkError = require('../errors/sdk-error');
var ClassificationResult = require('./classification-result');
var Nlu = require('./nlu');

/**
 * NLU using Botfuel Trainer API
 */

var BotfuelNlu = function (_Nlu) {
  _inherits(BotfuelNlu, _Nlu);

  /** @inheritdoc */
  function BotfuelNlu(config) {
    _classCallCheck(this, BotfuelNlu);

    logger.debug('constructor', config);

    var _this = _possibleConstructorReturn(this, (BotfuelNlu.__proto__ || Object.getPrototypeOf(BotfuelNlu)).call(this, config));

    _this.extractor = null;

    if (!process.env.BOTFUEL_APP_TOKEN) {
      throw new SdkError('BOTFUEL_APP_TOKEN is required for using the nlu service');
    }

    if (!process.env.BOTFUEL_APP_ID) {
      throw new SdkError('BOTFUEL_APP_ID is required for using the nlu service');
    }

    if (!process.env.BOTFUEL_APP_KEY) {
      throw new SdkError('BOTFUEL_APP_KEY is required for using the nlu service');
    }

    if (_this.config) {
      var classificationFilterPath = _this.config.path + '/src/classification-filter.js';
      if (fsExtra.pathExistsSync(classificationFilterPath)) {
        _this.classificationFilter = require(classificationFilterPath);
      }
    }
    return _this;
  }

  /**
   * Gets extractor files.
   * @param {String} path - extractors path
   * @returns {Array.<string>} - extractor files
   */


  _createClass(BotfuelNlu, [{
    key: 'getExtractorFiles',
    value: function getExtractorFiles(path) {
      var files = [];
      if (fs.existsSync(path)) {
        files = dir.files(path, { sync: true }) || files;
      }
      return files.filter(function (file) {
        return file.match(/^.*.js$/);
      });
    }

    /**
     * Gets extractors.
     * @param {String} path - extractors path
     * @returns {Array.<*>} - extractor instances
     */

  }, {
    key: 'getExtractors',
    value: function getExtractors(path) {
      // user extractors
      var extractors = this.getExtractorFiles(path).map(function (file) {
        var ExtractorConstructor = require(file);
        return new ExtractorConstructor(ExtractorConstructor.params);
      });
      // system extractors
      extractors.push(new BooleanExtractor({ locale: this.config.locale }));
      extractors.push(new LocationExtractor({}));
      return extractors;
    }

    /** @inheritdoc */

  }, {
    key: 'init',
    value: function () {
      var _ref = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee() {
        return regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                logger.debug('init');
                _get(BotfuelNlu.prototype.__proto__ || Object.getPrototypeOf(BotfuelNlu.prototype), 'init', this).call(this);

                // Extractors
                this.extractor = new CompositeExtractor({
                  extractors: this.getExtractors(this.config.path + '/src/extractors')
                });

              case 3:
              case 'end':
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function init() {
        return _ref.apply(this, arguments);
      }

      return init;
    }()

    /** @inheritdoc */

  }, {
    key: 'compute',
    value: function () {
      var _ref2 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee2(sentence, context) {
        var messageEntities, trainerUrl, options, res, classificationResults;
        return regeneratorRuntime.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                logger.debug('compute', sentence); // Context is not loggable

                // compute entities
                _context2.next = 3;
                return this.computeEntities(sentence);

              case 3:
                messageEntities = _context2.sent;


                // compute intents
                trainerUrl = process.env.BOTFUEL_TRAINER_API_URL || 'https://api.botfuel.io/trainer/api/v0';


                if (trainerUrl.slice(-1) !== '/') {
                  trainerUrl += '/';
                }

                options = {
                  uri: trainerUrl + 'classify',
                  qs: {
                    sentence: sentence
                  },
                  headers: {
                    'Botfuel-Bot-Id': process.env.BOTFUEL_APP_TOKEN,
                    'App-Id': process.env.BOTFUEL_APP_ID,
                    'App-Key': process.env.BOTFUEL_APP_KEY
                  },
                  json: true
                };
                _context2.next = 9;
                return rp(options);

              case 9:
                res = _context2.sent;
                classificationResults = res.map(function (data) {
                  return new ClassificationResult(data);
                });

                if (!this.classificationFilter) {
                  _context2.next = 16;
                  break;
                }

                _context2.next = 14;
                return this.classificationFilter(classificationResults, context);

              case 14:
                classificationResults = _context2.sent;

                classificationResults = classificationResults.slice(0, this.config.multiIntent ? 2 : 1);

              case 16:
                return _context2.abrupt('return', { messageEntities: messageEntities, classificationResults: classificationResults });

              case 17:
              case 'end':
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function compute(_x, _x2) {
        return _ref2.apply(this, arguments);
      }

      return compute;
    }()

    /**
     * Computes entities using the classifier.
     * @param {String} sentence - the user sentence
     * @returns {Object} entities
     */

  }, {
    key: 'computeEntities',
    value: function () {
      var _ref3 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee3(sentence) {
        var entities;
        return regeneratorRuntime.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                logger.debug('computeEntities', sentence);
                _context3.next = 3;
                return this.extractor.compute(sentence);

              case 3:
                entities = _context3.sent;
                return _context3.abrupt('return', entities);

              case 5:
              case 'end':
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function computeEntities(_x3) {
        return _ref3.apply(this, arguments);
      }

      return computeEntities;
    }()
  }]);

  return BotfuelNlu;
}(Nlu);

module.exports = BotfuelNlu;
