// botfuel-dialog/build/brains/brain.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

/**
 * Copyright (c) 2017 - present, Botfuel (https://www.botfuel.io).
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

var uuidv4 = require('uuid/v4');
var logger = require('logtown')('Brain');
var MissingImplementationError = require('../errors/missing-implementation-error');

/**
 * A brain is a storage for user and conversation data.
 * Some of the brain methods use a scope which is either 'user' or 'last conversation'.
 */

var Brain = function () {
  function Brain(config) {
    _classCallCheck(this, Brain);

    this.conversationDuration = config.brain.conversationDuration;
  }

  /**
   * Initializes the brain.
   * @private
   */


  _createClass(Brain, [{
    key: 'init',
    value: function () {
      var _ref = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee() {
        return regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                logger.debug('init');

              case 1:
              case 'end':
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function init() {
        return _ref.apply(this, arguments);
      }

      return init;
    }()

    /**
     * Empties the brain.
     * @abstract
     */

  }, {
    key: 'clean',
    value: function () {
      var _ref2 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee2() {
        return regeneratorRuntime.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                throw new MissingImplementationError();

              case 1:
              case 'end':
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function clean() {
        return _ref2.apply(this, arguments);
      }

      return clean;
    }()

    /**
     * Gets the init value for creating a new user.
     */

  }, {
    key: 'getUserInitValue',
    value: function getUserInitValue(userId) {
      return {
        _userId: userId,
        _conversations: [this.getConversationInitValue()],
        _createdAt: Date.now()
      };
    }

    /**
     * Adds a user if necessary (if he does not exist).
     */

  }, {
    key: 'addUserIfNecessary',
    value: function () {
      var _ref3 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee3(userId) {
        var userExists;
        return regeneratorRuntime.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                logger.debug('addUserIfNecessary', userId);
                _context3.next = 3;
                return this.hasUser(userId);

              case 3:
                userExists = _context3.sent;

                if (userExists) {
                  _context3.next = 7;
                  break;
                }

                _context3.next = 7;
                return this.addUser(userId);

              case 7:
              case 'end':
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function addUserIfNecessary(_x) {
        return _ref3.apply(this, arguments);
      }

      return addUserIfNecessary;
    }()

    /**
     * Checks if there is a user for a given id.
     * @abstract
     */

  }, {
    key: 'hasUser',
    value: function () {
      var _ref4 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee4(userId) {
        return regeneratorRuntime.wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                throw new MissingImplementationError();

              case 1:
              case 'end':
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));

      function hasUser(_x2) {
        return _ref4.apply(this, arguments);
      }

      return hasUser;
    }()

    /**
     * Adds a user.
     */

  }, {
    key: 'addUser',
    value: function () {
      var _ref5 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee5(userId) {
        return regeneratorRuntime.wrap(function _callee5$(_context5) {
          while (1) {
            switch (_context5.prev = _context5.next) {
              case 0:
                throw new MissingImplementationError();

              case 1:
              case 'end':
                return _context5.stop();
            }
          }
        }, _callee5, this);
      }));

      function addUser(_x3) {
        return _ref5.apply(this, arguments);
      }

      return addUser;
    }()

    /**
     * Gets a user.
     */

  }, {
    key: 'getUser',
    value: function () {
      var _ref6 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee6(userId) {
        return regeneratorRuntime.wrap(function _callee6$(_context6) {
          while (1) {
            switch (_context6.prev = _context6.next) {
              case 0:
                throw new MissingImplementationError();

              case 1:
              case 'end':
                return _context6.stop();
            }
          }
        }, _callee6, this);
      }));

      function getUser(_x4) {
        return _ref6.apply(this, arguments);
      }

      return getUser;
    }()

    /**
     * Gets all users.
     * @returns the users
     */

  }, {
    key: 'getAllUsers',
    value: function () {
      var _ref7 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee7() {
        return regeneratorRuntime.wrap(function _callee7$(_context7) {
          while (1) {
            switch (_context7.prev = _context7.next) {
              case 0:
                throw new MissingImplementationError();

              case 1:
              case 'end':
                return _context7.stop();
            }
          }
        }, _callee7, this);
      }));

      function getAllUsers() {
        return _ref7.apply(this, arguments);
      }

      return getAllUsers;
    }()

    /**
     * Gets the init value for creating a new conversation.
     */

  }, {
    key: 'getConversationInitValue',
    value: function getConversationInitValue() {
      return {
        _dialogs: {
          stack: [],
          previous: []
        },
        _createdAt: Date.now(),
        uuid: uuidv4()
      };
    }

    /**
     * Adds a conversation to a user.
     * @returns the last conversation added
     */

  }, {
    key: 'addConversation',
    value: function () {
      var _ref8 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee8(userId) {
        return regeneratorRuntime.wrap(function _callee8$(_context8) {
          while (1) {
            switch (_context8.prev = _context8.next) {
              case 0:
                throw new MissingImplementationError();

              case 1:
              case 'end':
                return _context8.stop();
            }
          }
        }, _callee8, this);
      }));

      function addConversation(_x5) {
        return _ref8.apply(this, arguments);
      }

      return addConversation;
    }()

    /**
     * Fetches the last conversation of the user.
     * If the last conversation found isn't valid anymore, adds a new one and returns it.
     * @returns the last conversation of the user
     */

  }, {
    key: 'fetchLastConversation',
    value: function () {
      var _ref9 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee9(userId) {
        return regeneratorRuntime.wrap(function _callee9$(_context9) {
          while (1) {
            switch (_context9.prev = _context9.next) {
              case 0:
                throw new MissingImplementationError();

              case 1:
              case 'end':
                return _context9.stop();
            }
          }
        }, _callee9, this);
      }));

      function fetchLastConversation(_x6) {
        return _ref9.apply(this, arguments);
      }

      return fetchLastConversation;
    }()

    /**
     * Sets a value for a key within the scope of the user.
     * @abstract
     * @returns the updated user
     */

  }, {
    key: 'userSet',
    value: function () {
      var _ref10 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee10(userId, key, value) {
        return regeneratorRuntime.wrap(function _callee10$(_context10) {
          while (1) {
            switch (_context10.prev = _context10.next) {
              case 0:
                throw new MissingImplementationError();

              case 1:
              case 'end':
                return _context10.stop();
            }
          }
        }, _callee10, this);
      }));

      function userSet(_x7, _x8, _x9) {
        return _ref10.apply(this, arguments);
      }

      return userSet;
    }()

    /**
     * Gets a value for a key within the scope of the user.
     */

  }, {
    key: 'userGet',
    value: function () {
      var _ref11 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee11(userId, key) {
        var user;
        return regeneratorRuntime.wrap(function _callee11$(_context11) {
          while (1) {
            switch (_context11.prev = _context11.next) {
              case 0:
                logger.debug('userGet', userId, key);
                _context11.next = 3;
                return this.getUser(userId);

              case 3:
                user = _context11.sent;
                return _context11.abrupt('return', user[key]);

              case 5:
              case 'end':
                return _context11.stop();
            }
          }
        }, _callee11, this);
      }));

      function userGet(_x10, _x11) {
        return _ref11.apply(this, arguments);
      }

      return userGet;
    }()

    /**
     * Sets a value for a key within the scope of the last conversation of a user.
     * @returns the updated conversation
     */

  }, {
    key: 'conversationSet',
    value: function () {
      var _ref12 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee12(userId, key, value) {
        return regeneratorRuntime.wrap(function _callee12$(_context12) {
          while (1) {
            switch (_context12.prev = _context12.next) {
              case 0:
                throw new MissingImplementationError();

              case 1:
              case 'end':
                return _context12.stop();
            }
          }
        }, _callee12, this);
      }));

      function conversationSet(_x12, _x13, _x14) {
        return _ref12.apply(this, arguments);
      }

      return conversationSet;
    }()

    /**
     * Gets the value for a given key within the scope of the last conversation of a user.
     * @returns the value
     */

  }, {
    key: 'conversationGet',
    value: function () {
      var _ref13 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee13(userId, key) {
        var conversation;
        return regeneratorRuntime.wrap(function _callee13$(_context13) {
          while (1) {
            switch (_context13.prev = _context13.next) {
              case 0:
                logger.debug('conversationGet', userId, key);
                _context13.next = 3;
                return this.fetchLastConversation(userId);

              case 3:
                conversation = _context13.sent;
                return _context13.abrupt('return', conversation[key]);

              case 5:
              case 'end':
                return _context13.stop();
            }
          }
        }, _callee13, this);
      }));

      function conversationGet(_x15, _x16) {
        return _ref13.apply(this, arguments);
      }

      return conversationGet;
    }()

    /**
     * Validates the last conversation of an user.
     */

  }, {
    key: 'isConversationValid',
    value: function isConversationValid(conversation) {
      return conversation !== undefined && Date.now() - conversation._createdAt < this.conversationDuration;
    }

    /**
     * Gets dialogs data from the last conversation.
     */

  }, {
    key: 'getDialogs',
    value: function () {
      var _ref14 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee14(userId) {
        var dialogs;
        return regeneratorRuntime.wrap(function _callee14$(_context14) {
          while (1) {
            switch (_context14.prev = _context14.next) {
              case 0:
                _context14.next = 2;
                return this.conversationGet(userId, '_dialogs');

              case 2:
                dialogs = _context14.sent;
                return _context14.abrupt('return', dialogs);

              case 4:
              case 'end':
                return _context14.stop();
            }
          }
        }, _callee14, this);
      }));

      function getDialogs(_x17) {
        return _ref14.apply(this, arguments);
      }

      return getDialogs;
    }()

    /**
     * Sets dialogs data in the last conversation.
     */

  }, {
    key: 'setDialogs',
    value: function () {
      var _ref15 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee15(userId, dialogs) {
        return regeneratorRuntime.wrap(function _callee15$(_context15) {
          while (1) {
            switch (_context15.prev = _context15.next) {
              case 0:
                if (!dialogs.isNewConversation) {
                  _context15.next = 4;
                  break;
                }

                _context15.next = 3;
                return this.addConversation(userId);

              case 3:
                delete dialogs.isNewConversation;

              case 4:
                _context15.next = 6;
                return this.conversationSet(userId, '_dialogs', dialogs);

              case 6:
              case 'end':
                return _context15.stop();
            }
          }
        }, _callee15, this);
      }));

      function setDialogs(_x18, _x19) {
        return _ref15.apply(this, arguments);
      }

      return setDialogs;
    }()

    /**
     * Gets a value for a key within the global scope.
     * @abstract
     */

  }, {
    key: 'botGet',
    value: function () {
      var _ref16 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee16(key) {
        return regeneratorRuntime.wrap(function _callee16$(_context16) {
          while (1) {
            switch (_context16.prev = _context16.next) {
              case 0:
                throw new MissingImplementationError();

              case 1:
              case 'end':
                return _context16.stop();
            }
          }
        }, _callee16, this);
      }));

      function botGet(_x20) {
        return _ref16.apply(this, arguments);
      }

      return botGet;
    }()

    /**
     * Sets a value for a key within the global scope.
     * @abstract
     */

  }, {
    key: 'botSet',
    value: function () {
      var _ref17 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee17(key, value) {
        return regeneratorRuntime.wrap(function _callee17$(_context17) {
          while (1) {
            switch (_context17.prev = _context17.next) {
              case 0:
                throw new MissingImplementationError();

              case 1:
              case 'end':
                return _context17.stop();
            }
          }
        }, _callee17, this);
      }));

      function botSet(_x21, _x22) {
        return _ref17.apply(this, arguments);
      }

      return botSet;
    }()
  }]);

  return Brain;
}();

module.exports = Brain;
