// botfuel-dialog/build/messages/card.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * Copyright (c) 2017 - present, Botfuel (https://www.botfuel.io).
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

var MessageError = require('../errors/message-error');
var Action = require('./action');
var Part = require('./part');

/**
 * A card part.
 */

var Card = function (_Part) {
  _inherits(Card, _Part);

  /**
   * @constructor
   * @param {String} title - the title
   * @param {String} imageUrl - the image url
   * @param {Object[]} actions - an array of actions
   */
  function Card(title, imageUrl, actions) {
    _classCallCheck(this, Card);

    var _this = _possibleConstructorReturn(this, (Card.__proto__ || Object.getPrototypeOf(Card)).call(this));

    _this.title = title;
    _this.imageUrl = imageUrl;
    _this.actions = actions;
    return _this;
  }

  /** @inheritDoc */


  _createClass(Card, [{
    key: 'toJson',
    value: function toJson() {
      // TODO : this is very Messenger specific, let's generalize it!
      return {
        title: this.title,
        image_url: this.imageUrl,
        buttons: this.actions.map(function (action) {
          return action.toJson();
        })
      };
    }

    /** @inheritDoc */

  }, {
    key: 'validate',
    value: function validate() {
      this.validateString('card', this.title);
      this.validateUrl('card', this.imageUrl);
      this.validateActions('card', this.actions);
    }

    /**
     * Validates that a value is an array of actions.
     * @param {String} name - the type of the object being validated
     * @param {*} actions - the value being validated
     * @returns {void}
     */

  }, {
    key: 'validateActions',
    value: function validateActions(name, actions) {
      this.validateArray(name, actions);
      var _iteratorNormalCompletion = true;
      var _didIteratorError = false;
      var _iteratorError = undefined;

      try {
        for (var _iterator = actions[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
          var action = _step.value;

          if (action instanceof Action) {
            action.validate();
          } else {
            throw new MessageError({
              name: 'card',
              message: 'Object \'' + JSON.stringify(action) + '\' should be of type Action\''
            });
          }
        }
      } catch (err) {
        _didIteratorError = true;
        _iteratorError = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion && _iterator.return) {
            _iterator.return();
          }
        } finally {
          if (_didIteratorError) {
            throw _iteratorError;
          }
        }
      }
    }
  }]);

  return Card;
}(Part);

module.exports = Card;
