// botfuel-dialog/build/messages/message.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var ValidObject = require('./valid-object');

/**
 * An abstract message.
 */
/**
 * Copyright (c) 2017 - present, Botfuel (https://www.botfuel.io).
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

var Message = function (_ValidObject) {
  _inherits(Message, _ValidObject);

  /**
   * @constructor
   * @param type - the message type
   * @param sender - the message sender, the bot or the user
   * @param value - the message value
   * @param options - the message options
   */
  function Message(type, sender, value, options) {
    _classCallCheck(this, Message);

    var _this = _possibleConstructorReturn(this, (Message.__proto__ || Object.getPrototypeOf(Message)).call(this));

    _this.type = type;
    _this.sender = sender;
    _this.value = value;
    _this.options = options;
    return _this;
  }

  /**
   * Converts a message to json and adds to it the bot and user ids.
   * @param userId - the user id
   * @returns the json message
   */


  _createClass(Message, [{
    key: 'toJson',
    value: function toJson(userId) {
      var jsonOutput = {
        type: this.type,
        sender: this.sender,
        user: userId,
        payload: {
          value: this.valueAsJson()
        }
      };

      if (this.options) {
        jsonOutput.payload.options = this.options;
      }

      return jsonOutput;
    }

    /**
     * Returns the value as json.
     * @returns the json value
     */

  }, {
    key: 'valueAsJson',
    value: function valueAsJson() {
      return this.value;
    }
  }, {
    key: 'validate',
    value: function validate() {
      this.validateString(this.type, this.type);
      this.validateString(this.type, this.sender);
    }
  }]);

  return Message;
}(ValidObject);

module.exports = Message;
