// botfuel-dialog/build/resolver.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var fs = require('fs'); /**
                         * Copyright (c) 2017 - present, Botfuel (https://www.botfuel.io).
                         *
                         * Licensed under the Apache License, Version 2.0 (the "License");
                         * you may not use this file except in compliance with the License.
                         * You may obtain a copy of the License at
                         *
                         * http://www.apache.org/licenses/LICENSE-2.0
                         *
                         * Unless required by applicable law or agreed to in writing, software
                         * distributed under the License is distributed on an "AS IS" BASIS,
                         * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
                         * See the License for the specific language governing permissions and
                         * limitations under the License.
                         */

var path = require('path');
var logger = require('logtown')('Resolver');
var ResolutionError = require('./errors/resolution-error');
var MissingImplementationError = require('./errors/missing-implementation-error');

/**
 * The adapter resolver resolves the adapter at startup.
 */

var Resolver = function () {

  /**
   * @constructor
   * @param config - the bot config
   * @param {String} kind - the kind of objects we want to resolve
   */
  function Resolver(config, kind) {
    _classCallCheck(this, Resolver);

    this.config = config;
    this.kind = kind;
    this.directories = config.componentRoots.map(function (componentRoot) {
      return path.join(componentRoot, kind + 's');
    }).filter(fs.existsSync);
  }

  /**
   * Gets the possible filenames for a given name.
   * @param name - the component name
   * @returns the possible paths
   */


  _createClass(Resolver, [{
    key: 'getFilenames',
    value: function getFilenames(name) {
      return [name + '-' + this.kind + '.js'];
    }

    /**
     * Gets the possible paths for a given name.
     * @param name - the component name
     * @returns the possible paths
     */

  }, {
    key: 'getPaths',
    value: function getPaths(name) {
      logger.debug('getPaths', name);

      var possibleFilenames = this.getFilenames(name);

      return this.directories.map(function (directory) {
        return possibleFilenames.map(function (filename) {
          return path.join(directory, filename);
        });
      }).reduce(function (a, b) {
        return [].concat(_toConsumableArray(a), _toConsumableArray(b));
      });
    }

    /**
     * Gets the path for a given name.
     * @param name - the component name
     * @returns the path if exists or null
     */

  }, {
    key: 'getPath',
    value: function getPath(name) {
      logger.debug('getPath');
      var _iteratorNormalCompletion = true;
      var _didIteratorError = false;
      var _iteratorError = undefined;

      try {
        for (var _iterator = this.getPaths(name)[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
          var componentPath = _step.value;

          logger.debug('getPath: test path', componentPath);
          if (fs.existsSync(componentPath)) {
            logger.debug('getPath: existing path', componentPath);
            return componentPath;
          }
        }
      } catch (err) {
        _didIteratorError = true;
        _iteratorError = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion && _iterator.return) {
            _iterator.return();
          }
        } finally {
          if (_didIteratorError) {
            throw _iteratorError;
          }
        }
      }

      return null;
    }

    /**
     * Resolves the component for a given name.
     * @param name - the component name
     * @returns the component instance or null
     */

  }, {
    key: 'resolve',
    value: function resolve(name) {
      logger.debug('resolve', name);
      var componentPath = this.getPath(name);
      if (componentPath) {
        var Resolved = require(componentPath);
        return this.resolutionSucceeded(Resolved);
      }
      throw new ResolutionError({
        kind: this.kind,
        name: name,
        paths: this.getPaths(name)
      });
    }

    /**
     * Instantiates a component after a successful resolution
     * @param Resolved - the component class
     * @returns the instance
     */

  }, {
    key: 'resolutionSucceeded',
    value: function resolutionSucceeded(componentClass) {
      throw new MissingImplementationError();
    }
  }]);

  return Resolver;
}();

module.exports = Resolver;
