// botfuel-dialog/build/adapters/web-adapter.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var url = require('url'); /**
                           * Copyright (c) 2017 - present, Botfuel (https://www.botfuel.io).
                           *
                           * Licensed under the Apache License, Version 2.0 (the "License");
                           * you may not use this file except in compliance with the License.
                           * You may obtain a copy of the License at
                           *
                           * http://www.apache.org/licenses/LICENSE-2.0
                           *
                           * Unless required by applicable law or agreed to in writing, software
                           * distributed under the License is distributed on an "AS IS" BASIS,
                           * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
                           * See the License for the specific language governing permissions and
                           * limitations under the License.
                           */

var querystring = require('querystring');
var express = require('express');
var exphbs = require('express-handlebars');
var rp = require('request-promise-native');
var bodyParser = require('body-parser');
var logger = require('logtown')('WebAdapter');
var MissingImplementationError = require('../errors/missing-implementation-error');
var Adapter = require('./adapter');

// absolute urls to static and template folders
var PORT = process.env.PORT || process.env.BOTFUEL_ADAPTER_PORT || 5000;
var BOT_URL = process.env.BOT_URL || 'http://localhost:' + PORT;
var STATIC_BASE_URL = url.resolve(BOT_URL, 'static/');
var TEMPLATE_BASE_URL = url.resolve(BOT_URL, 'templates/');

// screenshot service url
var SCREENSHOT_SERVICE_URL = 'https://botfuel-screenshot-service.herokuapp.com/';

/**
 * Generic web adapter (to be subclassed), it serves:
 * - /webhook : requests to the bot
 * - /static: static files under src/static
 * - /templates: handlebars templates under src/templates
 * @extends Adapter
 */

var WebAdapter = function (_Adapter) {
  _inherits(WebAdapter, _Adapter);

  function WebAdapter() {
    _classCallCheck(this, WebAdapter);

    return _possibleConstructorReturn(this, (WebAdapter.__proto__ || Object.getPrototypeOf(WebAdapter)).apply(this, arguments));
  }

  _createClass(WebAdapter, [{
    key: 'run',

    /** @inheritDoc */
    value: function () {
      var _ref = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee() {
        var app, hbs;
        return regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                logger.debug('run');
                app = express();

                app.use(bodyParser.json());
                app.use('/static', express.static('src/static'));
                hbs = exphbs.create({});

                app.engine('handlebars', hbs.engine);
                app.set('view engine', 'handlebars');
                app.set('views', 'src/templates');
                this.createRoutes(app);
                app.listen(PORT, function () {
                  return logger.info('run: listening on port', PORT);
                });

              case 10:
              case 'end':
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function run() {
        return _ref.apply(this, arguments);
      }

      return run;
    }()

    /**
     * Creates routes.
     * @param {Object} app - the express app
     * @returns {void}
     */

  }, {
    key: 'createRoutes',
    value: function createRoutes(app) {
      var _this2 = this;

      logger.debug('createRoutes');
      app.post('/webhook', function (req, res) {
        return _this2.handleRequest(req, res);
      });
      app.get('/templates/:id', function (req, res) {
        return _this2.handleTemplate(req, res);
      });
    }

    /**
     * Webhook used for handling requests to the bot.
     * @param req - the request object
     * @param res - the response object
     */

  }, {
    key: 'handleRequest',
    value: function () {
      var _ref2 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee2(req, res) {
        return regeneratorRuntime.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                throw new MissingImplementationError();

              case 1:
              case 'end':
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function handleRequest(_x, _x2) {
        return _ref2.apply(this, arguments);
      }

      return handleRequest;
    }()

    /**
     * Webhook used for handling template requests.
     * @param req - the request object
     * @param res - the response object
     */

  }, {
    key: 'handleTemplate',
    value: function () {
      var _ref3 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee3(req, res) {
        var templateParameters;
        return regeneratorRuntime.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                logger.debug('handleTemplate', { id: req.params.id, query: req.query });

                templateParameters = req.query;

                res.render(req.params.id, templateParameters, function (err, html) {
                  if (err) {
                    logger.error('Could not render the handlebars template: ' + req.params.id);
                    res.status(400).send(err);
                  } else {
                    res.status(200).send(html);
                  }
                });

              case 3:
              case 'end':
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function handleTemplate(_x3, _x4) {
        return _ref3.apply(this, arguments);
      }

      return handleTemplate;
    }()

    /** @inheritDoc */

  }, {
    key: 'sendMessage',
    value: function () {
      var _ref4 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee4(botMessage) {
        var requestOptions, res;
        return regeneratorRuntime.wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                requestOptions = {
                  uri: this.getUrl(botMessage),
                  qs: this.getQueryParameters(botMessage),
                  body: this.getBody(botMessage),
                  json: true
                };
                _context4.prev = 1;
                _context4.next = 4;
                return rp.post(requestOptions);

              case 4:
                res = _context4.sent;

                if (res.statusCode && res.statusCode !== 200) {
                  // not handled on messenger
                  logger.error('postResponse: KO', res);
                } else {
                  logger.debug('postResponse: OK');
                }
                _context4.next = 11;
                break;

              case 8:
                _context4.prev = 8;
                _context4.t0 = _context4['catch'](1);

                // TODO: is this what we want?
                logger.error('postResponse: catch KO', _context4.t0.message || _context4.t0.error || _context4.t0);

              case 11:
              case 'end':
                return _context4.stop();
            }
          }
        }, _callee4, this, [[1, 8]]);
      }));

      function sendMessage(_x5) {
        return _ref4.apply(this, arguments);
      }

      return sendMessage;
    }()

    /**
     * Get the URL of the API used to send a bot message.
     * @param botMessage - the bot response
     */

  }, {
    key: 'getUrl',
    value: function getUrl(botMessage) {
      // eslint-disable-line no-unused-vars
      throw new MissingImplementationError();
    }

    /**
     * Get the query parameters to send along with a bot message.
     * @param botMessage - the bot response
     */

  }, {
    key: 'getQueryParameters',
    value: function getQueryParameters(botMessage) {
      // eslint-disable-line no-unused-vars
      throw new MissingImplementationError();
    }

    /**
     * Get the body of the request when sending a bot message.
     * @param botMessage - the bot response
     */

  }, {
    key: 'getBody',
    value: function getBody(botMessage) {
      // eslint-disable-line no-unused-vars
      throw new MissingImplementationError();
    }

    /**
     * Get absolute url for resource inside static folder
     * @param {String} resourcePath - resource path relative to static, ex: images/logo.png
     * @returns {null}
     */

  }], [{
    key: 'getStaticUrl',
    value: function getStaticUrl(resourcePath) {
      logger.debug('getStaticUrl', resourcePath);
      return url.resolve(STATIC_BASE_URL, resourcePath);
    }

    /**
     * Get absolute path for template
     * @param templateName - handlebars template name
     * @param params - parameters to be passed to the template
     */

  }, {
    key: 'getTemplateUrl',
    value: function getTemplateUrl(templateName, params) {
      var templateRoot = url.resolve(TEMPLATE_BASE_URL, templateName);
      var templateUrl = templateRoot + '?' + querystring.stringify(params);
      logger.debug('getTemplateUrl', templateUrl);

      return templateUrl;
    }

    /**
     * Get absolute path for the screenshot template image
     * @async
     * @param {String} templateName - handlebars template name
     * @param {Object} params - parameters to be passed to the template
     * @param {Number} width - image width
     * @param {Number} height - image height
     * @param {Number} quality - image quality
     */

  }, {
    key: 'getImageUrl',
    value: function getImageUrl(templateName, params) {
      var width = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 800;
      var height = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : 600;
      var quality = arguments.length > 4 && arguments[4] !== undefined ? arguments[4] : 80;

      var templateUrl = this.getTemplateUrl(templateName, params);
      var screenshotParams = {
        url: templateUrl,
        quality: quality,
        width: width,
        height: height
      };

      return SCREENSHOT_SERVICE_URL + '?' + querystring.stringify(screenshotParams);
    }
  }]);

  return WebAdapter;
}(Adapter);

module.exports = WebAdapter;
